### PR TITLE
D3ASIM-3486: Fixed: Don't send results when run via cli.

### PR DESCRIPTION
**Pull request by hannesdiedrich**

### CODE REVIEW

**Review by Hassan754 — DISMISSED** [review]
LGTM

File: tests/test_simulation.py
Comment (fievelk) [inline tests/test_simulation.py:61]:
```suggestion
    def test_results_are_sent_via_kafka_if_not_started_from_cli():
```

File: tests/test_simulation.py
Comment (fievelk) [inline tests/test_simulation.py:34]:
```suggestion
    def test_results_not_sent_via_kafka_if_started_from_cli():
```

Comment (fievelk) [reply tests/test_simulation.py:34]: Extremely minor: I just suggested changing the spelling from `send` to `sent`.

File: tests/test_simulation.py
Comment (fievelk) [inline tests/test_simulation.py:53]: If you want to follow PEP-8, you can avoid using the backslash (which is discouraged) and do this:
```suggestion
        simulation.endpoint_buffer.results_handler.all_ui_results = {
            k: {} for k in results_mapping}
```
I don't know why PyCharm doesn't enforce this yet. They have an open ticket for it anyway (https://youtrack.jetbrains.com/issue/PY-17193).

**Review by fievelk — DISMISSED** [review]
LGTM. Only very minor comments that can be ignored.

File: src/d3a/d3a_core/simulation.py
Comment (faizan2590) [inline src/d3a/d3a_core/simulation.py:269]:
```suggestion
        if not self._started_from_cli and self.kafka_connection.is_enabled:
```
How about making `is_enabled` a property method?

**Review by faizan2590 — DISMISSED** [review]
One suggestion, other than that LGTM

File: src/d3a/d3a_core/simulation.py
Comment (spyrostz) [inline src/d3a/d3a_core/simulation.py:269]: If you want to go this way, then you should add the self._started_from_cli to the should_export_stats method, in order to be applied for the whole simulation class and not just here.

Comment (spyrostz) [reply src/d3a/d3a_core/simulation.py:269]: I do not really recall why, but I think that I stumbled upon issues while trying to implement exactly this (probably with unit tests and mocking the property IIRC )

**Review by spyrostz — DISMISSED** [review]
LGTM, but FYI, same as with Redis before introducing Kafka, you do not really need this, since disabling Kafka should allow the CLI to run (have a look at `should_export_results` as discussed in the comment too). Redis was also sending the results in the past if one had a running redis server while running via CLI, the most probably reason why it did not crash was because the results were not validated.
Having said that, it is okay to have this as an extra precaution

**Comment by hannesdiedrich** [comment]
> LGTM, but FYI, same as with Redis before introducing Kafka, you do not really need this, since disabling Kafka should allow the CLI to run (have a look at `should_export_results` as discussed in the comment too). Redis was also sending the results in the past if one had a running redis server while running via CLI, the most probably reason why it did not crash was because the results were not validated.
> Having said that, it is okay to have this as an extra precaution

@spyrostz I had a look even before and the results where indeed validated before your change, have a look here: https://github.com/gridsingularity/d3a/pull/1137/files#diff-78571cbd782c75f4e8975fac86e013383e9b02c5c3bc1b63f7889a6c1c91423cL176
I thought the early return of `publish _results` when redis is not enabled was the difference. But you are right, if a redis server was running, we would have the same issue. But then I can not really understand why this fails locally. 
I would keep this change, as you agreed that it does not make sense to send results via kafka if you run locally.

File: src/d3a/d3a_core/simulation.py
Comment (hannesdiedrich) [reply src/d3a/d3a_core/simulation.py:269]: OK, great! Will change that.

Comment (hannesdiedrich) [reply src/d3a/d3a_core/simulation.py:269]: Actually, now that I tried it again (had before but forgot). 
We actually look for the case here: `KAFKA=True` + `LOCAL_CLI_RUN=False`.
Your proposal of `self.should_export_plots` would be: `KAFKA=False` + `LOCAL_CLI_RUN=True`
which is indeed the invert of the above.
However, we like to omit sending results for the case: `KAFKA=True` + `LOCAL_CLI_RUN=True`

Comment (spyrostz) [reply src/d3a/d3a_core/simulation.py:269]: My proposal was to actually use an or condition and not the and condition that you mentioned:
`return self._started_from_cli or not self.kafka_connection.is_enabled()`

Comment (hannesdiedrich) [reply src/d3a/d3a_core/simulation.py:269]: Ill have an idea how to fix it. Will propose something.

File: src/d3a/d3a_core/simulation.py
Comment (fievelk) [inline src/d3a/d3a_core/simulation.py:409]: Nit: the naming/documentation seems to be still a bit unclear to me (where should the results be sent? Why?)

**Review by fievelk — DISMISSED** [review]
LGTM

File: src/d3a/d3a_core/simulation.py
Comment (spyrostz) [reply src/d3a/d3a_core/simulation.py:409]: I think that should_send_results_to_broker would help clarify this. I would suggest should_publish_result if we sticked to Redis, however we are not running on pubsub anymore, therefore it is a misleading name.

File: integration_tests/steps/integration_tests.py
Comment (spyrostz) [inline integration_tests/steps/integration_tests.py:605]: Reminder to remove before merging

File: src/d3a/d3a_core/simulation.py
Comment (spyrostz) [inline src/d3a/d3a_core/simulation.py:271]: I would further refactor all this to the following structure:
`if self.export_results_on_finish: .... `
`elif self.should_send_results: ....`
Reason for that is because these statements are mutually exclusive, therefore we are certain that if we are exporting the simulation results we should not send them via the message broker.

**Review by spyrostz — APPROVED** [review]
LGTM once comments are dealt with

**Review by fievelk — APPROVED** [review]

**Review by spyrostz — APPROVED** [review]
LGTM

**Pull request merged**
